clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
-static bool isZeroSizedArray(QualType Ty) {
-  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
-    return CATy->getSize() == 0;
+static bool isZeroSizedArray(Sema &SemaRef, QualType Ty) {
+  if (const auto *CAT = SemaRef.getASTContext().getAsConstantArrayType(Ty))
+    return CAT->getSize() == 0;
   return false;
 }
 
+// NOTE: In case this function gets upstreamed and other single source
+// offloading models also need it, it can be generalized via using of
+// `targetDiag` instead of `SYCLDiagIfDeviceCode` and removing assertion at the
+// beginning.
+void Sema::deepTypeCheckForSYCLDevice(SourceLocation UsedAt,
```

Comment:
This is a nit but I think renaming this to `CheckArrayType `or something along that is better . Or are you naming it in a generic manner to future proof it in case you need this check for other types?

Reply:
> Or are you naming it in a generic manner to future proof it in case you need this check for other types?

Yeah.